metrics/providers/micrometer/src/main/java/io/helidon/metrics/providers/micrometer/MMeterRegistry.java

```diff
@@ -215,39 +221,65 @@ static MMeterRegistry applyMetersProvidersToRegistry(MetricsFactory factory,
 
     @Override
     public void close() {
-        onAddListeners.clear();
-        onRemoveListeners.clear();
-        List.copyOf(meters.values()).forEach(this::remove);
-        meters.clear();
-        buildersByPromMeterId.clear();
-        scopeMembership.clear();
-        metersById.clear();
+        lock.writeLock().lock();
+        try {
+            onAddListeners.clear();
+            onRemoveListeners.clear();
+            List.copyOf(meters.values()).forEach(this::remove);
+            meters.clear();
+            buildersByPromMeterId.clear();
+            scopeMembership.clear();
+            metersById.clear();
+        } finally {
+            lock.writeLock().unlock();
+        }
     }
 
     @Override
     public List<io.helidon.metrics.api.Meter> meters() {
-        return meters.values()
-                .stream()
+        lock.readLock().lock();
+        List<io.helidon.metrics.api.Meter> temp;
+        try {
+            temp = new ArrayList<>(meters.values());
+        } finally {
+            lock.readLock().unlock();
+        }
+        return temp.stream()
                 .map(io.helidon.metrics.api.Meter.class::cast)
                 .toList();
     }
 
     @Override
     public Collection<io.helidon.metrics.api.Meter> meters(Predicate<io.helidon.metrics.api.Meter> filter) {
-        return meters.values()
-                .stream()
+        lock.readLock().lock();
+        List<io.helidon.metrics.api.Meter> temp;
+        try {
+            temp = new ArrayList<>(meters.values());
+        } finally {
+            lock.readLock().unlock();
+        }
+        return temp.stream()
                 .map(io.helidon.metrics.api.Meter.class::cast)
                 .filter(filter)
                 .toList();
+
     }
 
     @Override
     public Iterable<String> scopes() {
-        return scopeMembership.keySet();
+        lock.readLock().lock();
+        try {
+            return new HashSet<>(scopeMembership.keySet());
```

Comment:
Is it OK if someone casts the returned `Iterable` to a `Set` and starts adding things to it? If not, consider:
```
return Set.copyOf(scopeMembership.keySet());
```

Reply:
We've copied the elements into a new set that is not backed by the original, so the calling code can do what it wants with the returned set without disturbing our internal set.

Or did I miss your point?

Reply:
I just meant: what are the semantics of the return value? I tend to prefer: "the return value is immutable". You're saying, I think: "the return value's mutability is undefined". Either can be true; just wanted clarification.